src/main/java/build/buildfarm/worker/CASFileCache.java

```diff
@@ -0,0 +1,238 @@
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.TimeUnit;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+class CASFileCache {
+  private final Instance instance;
+  private final Path root;
+  private final long maxSizeInBytes;
+  private final Map<Path, FileEntry> paths;
+  private final List<Path> accessScanPaths;
+
+  private long sizeInBytes;
+  private Queue<FileEntry> expirationQueue;
+
+  private static final class FileEntry implements Comparable<FileEntry> {
+    private final Path path;
+    private final long sizeInBytes;
+    private final long createdTime;
+    private long accessTime;
+
+    public FileEntry(
+        Path path, long sizeInBytes, long createdTime, long accessTime) {
+      this.path = path;
+      this.sizeInBytes = sizeInBytes;
+      this.createdTime = createdTime;
+      this.accessTime = accessTime;
+    }
+
+    public Path getPath() {
+      return path;
+    }
+
+    public long getSizeInBytes() {
+      return sizeInBytes;
+    }
+
+    public long getCreatedTime() {
+      return createdTime;
+    }
+
+    public long getAccessTime() {
+      return accessTime;
+    }
+
+    public void setAccessTime( long accessTime ) {
+      this.accessTime = accessTime;
+    }
+
+    @Override
+    public int compareTo(FileEntry entry) {
+      // the most recently accessed file is greater
+      if (accessTime != entry.accessTime) {
+        return entry.accessTime > accessTime ? 1 : -1;
+      }
+      // the smaller file is greater
+      if (sizeInBytes != entry.sizeInBytes) {
+        return entry.sizeInBytes > sizeInBytes ? 1 : -1;
+      }
+      // the most recently created file is greater
+      if (createdTime != entry.createdTime) {
+        return entry.createdTime > createdTime ? 1 : -1;
+      }
+      // path tie breaker
+      return path.compareTo(entry.path);
+    }
+  }
+
+  public CASFileCache(Instance instance, Path root, long maxSizeInBytes) {
+    this.instance = instance;
+    this.root = root;
+    this.maxSizeInBytes = maxSizeInBytes;
+    paths = new HashMap<>();
+    accessScanPaths = new ArrayList<>();
+    sizeInBytes = 0;
+  }
+
+  public void start() throws IOException {
+    Files.createDirectories(root);
+
+    Files.walkFileTree(root, new SimpleFileVisitor<Path>() {
+      @Override
+      public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+        long size = attrs.size();
+        if (sizeInBytes + size > maxSizeInBytes) {
+          // FIXME need to do this according to expiration strategy
+          Files.delete(file);
+        } else {
+          paths.put(file, new FileEntry(
+              file,
+              size,
+              attrs.lastModifiedTime().to(TimeUnit.NANOSECONDS),
+              attrs.lastAccessTime().to(TimeUnit.NANOSECONDS)));
+          sizeInBytes += size;
+        }
+        return FileVisitResult.CONTINUE;
+      }
+    });
+    update();
+  }
+
+  public void update() {
```

Comment:
Here also.

Reply:
Dropped this in favor of reference counters, hope that name is documentation enough.